### ISSUE TITLE
Improve dot performance on data that isn't contiguous

### ISSUE THREAD

**Issue by akhmerov** [numpy/numpy#19650]
Numpy dot product can experience a dramatic slowdown on non-contiguous data:

```python
import numpy as np
print(np.__version__)
# 1.21.1
a = np.random.randn(1000, 100, 1000)
%timeit a[:, 0, :] @ a[:, :, 20]
# 1.59 s ± 184 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit np.ascontiguousarray(a[:, 0, :]) @ np.ascontiguousarray(a[:, :, 20])
# 8.06 ms ± 410 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
While the workaround shown above is straightforward, the performance regression is hard to detect, and many users won't even be aware about it. Therefore I propose to improve the dot product to identify when data is highly discontinuous and copy it before performing the multiplication.

**Comment by seberg** [comment]
Well, the part of the ufunc refactor to make it use the `ArrayMethod` concept and and expand the core signature is just merged.  That should give anyone with the curiosity (and a reasonable amount of free time, unfortunately) the tools to do this (reasonably) well...  It probably isn't quite possible to do everything that may make sense, but at least a good deal (and this part).

**Comment by akhmerov** [comment]
Absolutely: I realize that the proper heuristic for when to copy is fairly complicated, just wanted to report this for a reference.

**Comment by thisisamardeep** [comment]
i am picking this up.Will update asap by end of sep.

**Comment by thisisamardeep** [comment]
Please assign this to me.

**Comment by seberg** [comment]
@thisisamardeep we don't have a habit of assigning issues, especially those that are a bit open ended.  Someone else finding the issue should ask here if you are still working on it instead.
I will note that this is in C-code, and may require a fair amount (or even a larger amount) of understanding the NumPy ufunc machinery.  It is probably doable, but is project sized, especially for a new contributor.

**Comment by thisisamardeep** [comment]
@seberg i have started digging into the code.it is helping me.i have a question.if you can outline some broad thoughts around what should or should not be the hueristics for determining when the data is highly discontiguos and copy it .i will work on those lines.i debugged the above snippet since the data is adjacent in memory so calculation is fast in second option.so basically we just need some logic which tells at what point we decide that  we want to copy before doing the actual calculation.i mean what is the measure of discontiuity for us to go for second option. i am typing from mobile.sorry for typos.

**Comment by seberg** [comment]
@thisisamardeep probably some size-based heuristic after timing a bit with openblas.  I honestly do not know.  The main difficulty is that it would likely make sense to make this a new `ArrayMethod` API so that the allocated work-space and heuristic decision can be reused over multiple calls to the inner-loop.

**Comment by thisisamardeep** [comment]
ok i will come up with somethimg over the weekend and then based on your feedback will improve the logic.

**Comment by eendebakpt** [comment]
This has been addressed in #23752. Closing the issue.

For int type the performance could also be improved with a similar approach.
